lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -122,69 +122,240 @@ const _: () = {
 };
 
 impl Client {
-    fn new(
-        iron_rdp_client: IronRDPClient,
+    pub fn connect(
         go_ref: usize,
         tokio_rt: tokio::runtime::Runtime,
-    ) -> *const Self {
-        Box::into_raw(Box::new(Self {
-            iron_rdp_client: Mutex::new(iron_rdp_client),
-            tokio_rt,
-            go_ref,
-        }))
+        params: ConnectParams,
+    ) -> Result<*const Self, ConnectError> {
```

Comment:
I'd expect this to just return a `Result<Self, ConnectError>` and the FFI wrapper can deal with boxing and returning it as a pointer, to move the actual implementation away from FFI details.

Reply:
I don't agree, the `Client` is fundamentally an "FFI thing" that lives life as a `*const Self`, it's created to be passed back over the FFI and has the methods `from_raw` and `drop` which (more or less) take this "FFI thing" as a receiver

```rust
/// # Safety
    ///
    /// client_ptr MUST be a valid pointer.
    /// (validity defined by https://doc.rust-lang.org/nightly/core/primitive.pointer.html#method.as_ref-1)
    pub unsafe fn from_raw<'a>(client_ptr: *const Self) -> Result<&'a Client, CGOErrCode> {
        client_ptr.as_ref().ok_or_else(|| {
            error!("Client pointer is null");
            CGOErrCode::ErrCodeClientPtr
        })
    }

    /// # Safety
    ///
    /// Calling this twice on the same Client
    /// pointer will result in a double free.
    pub unsafe fn drop(ptr: *mut Self) {
        if !ptr.is_null() {
            drop(Box::from_raw(ptr))
        }
    }
```

That said, it looks like we may be able to get rid of this `*const` grossness at some point, see https://github.com/gravitational/teleport/pull/29764#discussion_r1282416270